src/main/java/org/broadinstitute/hellbender/dev/tools/walkers/bqsr/ApplyBQSRDataflow.java

```diff
@@ -96,10 +107,43 @@ protected void setupPipeline(Pipeline pipeline) {
         final SAMSequenceDictionary sequenceDictionary = header.getSequenceDictionary();
         final List<SimpleInterval> intervals = intervalArgumentCollection.intervalsSpecified() ? intervalArgumentCollection.getIntervals(sequenceDictionary) :
                 IntervalUtils.getAllIntervalsForReference(sequenceDictionary);
-        final BaseRecalOutput recalInfo = new BaseRecalOutput(BQSR_RECAL_FILE);
-        PCollection<BaseRecalOutput> recalInfoSingletonCollection = pipeline.apply(Create.of(recalInfo).withName("recal_file ingest"));
+        PCollection<BaseRecalOutput> recalInfoSingletonCollection;
+        if (BucketUtils.isCloudStorageUrl(BQSR_RECAL_FILE_NAME)) {
+            recalInfoSingletonCollection = BaseRecalOutputSource.of(pipeline, BQSR_RECAL_FILE_NAME);
+        } else{
+            final BaseRecalOutput recalInfo = new BaseRecalOutput(new File(BQSR_RECAL_FILE_NAME));
+            recalInfoSingletonCollection = pipeline.apply(Create.of(recalInfo).withName("recal_file ingest"));
+        }
         PCollection<Read> output = readsSource.getReadPCollection(intervals, ValidationStringency.SILENT)
                 .apply(new ApplyBQSRTransform(header, recalInfoSingletonCollection, bqsrOpts));
-        SmallBamWriter.writeToFile(pipeline, output, header, OUTPUT);
+        intermediateGCSBam = OUTPUT;
+        if (isRemote() && !BucketUtils.isCloudStorageUrl(intermediateGCSBam)) {
+            // The user specified remote execution and provided a local file name. So we're going to have to save to GCS as a go-between.
+            // Note that this may require more permissions
+            copyIntermediate = true;
+            intermediateGCSBam = GcsPath.fromUri(stagingLocation).resolve("temp-applyBqsr-output-" + new Random().nextLong() + ".bam").toString();
+            logger.info("Staging results at " + intermediateGCSBam);
+        }
+        SmallBamWriter.writeToFile(pipeline, output, header, intermediateGCSBam);
+    }
+
+    @Override
+    protected void afterPipeline(Pipeline pipeline) {
+        if (!copyIntermediate) return;
+        try {
+            logger.info("Copying results from " + intermediateGCSBam + " to " + OUTPUT + ".");
+            BucketUtils.copyFile(intermediateGCSBam, pipeline.getOptions(), OUTPUT);
+        } catch (IOException x) {
+            // keep the intermediate file if anything goes wrong, so we can investigate
+            throw new UserException.CouldNotCreateOutputFile("Error writing to '" + OUTPUT + "'.",x);
+        }
+        try {
+            BucketUtils.deleteFile(intermediateGCSBam, pipeline.getOptions());
+        } catch (Exception x) {
```

Comment:
I could, but why? The function's signature guarantees we can't have any other checked exception type. And if a RuntimeException is thrown, do we really want to propagate it and kill the whole pipeline? It seems that leaving the temporary file around is preferable to tearing down the computation.